ansible_base/models/resource.py

```diff
@@ -0,0 +1,72 @@
+import uuid
+
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+
+from .service_id import service_id
+
+
+def short_service_id():
+    return service_id().split('-')[0]
+
+
+class ResourceType(models.Model):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        from ansible_base.resource_registry import get_registry
+
+        self.resource_registry = get_registry()
+
+    content_type = models.OneToOneField(ContentType, on_delete=models.CASCADE, related_name="resource_type", unique=True)
+    externally_managed = models.BooleanField()
+    migrated = models.BooleanField(null=False, default=False)
+    name = models.CharField(max_length=256, unique=True, db_index=True, editable=False, blank=False, null=False)
+
+    @property
+    def serializer_class(self):
+        return self.get_resource_config().managed_serializer
+
+    def get_resource_config(self):
+        return self.resource_registry.get_config_for_model(model=self.content_type.model_class())
+
+
+class Resource(models.Model):
+    # objects = ResourceManager()
+
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE, related_name="resources")
+
+    # this has to accommodate integer and UUID object IDs
+    object_id = models.TextField(null=False)
+    content_object = GenericForeignKey('content_type', 'object_id')
+
+    service_id = models.CharField(default=short_service_id, max_length=8)
+
+    # we're not using this as the primary key because the ansible_id can change if the object is
+    # externally managed.
+    resource_id = models.UUIDField(default=uuid.uuid4, db_index=True, unique=True)
+
+    # human readable name for the resource
+    name = models.CharField(max_length=256, null=True)
```

Comment:
The common `name` in AWX has max length of 512

https://github.com/ansible/awx/blob/30d4df788ff8d4be4fd4e503854cd432e1490586/awx/main/models/base.py#L358

same with Gateway I think. I think it's safe to just go with that.

Reply:
Should we abstract this into a constant in DAB that this and the other services can reference so we all stay in-sync?

Reply:
A common named model already exists in DAB. If inheriting isn't your cup of tea then there are ways to get reuse. I don't feel too particular about any of this, other than that this field can contain names from everything else.

https://github.com/ansible/django-ansible-base/blob/cdf5fb0e6879bd268f1f29926466cc631276af69/ansible_base/models/common.py#L111

Reply:
Actually, this is a good point, not everything inherits form NamedCommonModel in the first place. What if we needed/wanted to share a resource without a specific name? i.e. a token or an authenticator_map. Or are we making a hard requirement that anything that will be shared has to have a name field?

Reply:
Updated the max length to 512 and set it to truncate anything larger than that.